core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -74,6 +74,12 @@ class BlockManagerMasterEndpoint(
   // Mapping from block id to the set of block managers that have the block.
   private val blockLocations = new JHashMap[BlockId, mutable.HashSet[BlockManagerId]]
 
+  // Mapping from host name to shuffle (mergers) services
+  private val mergerLocations = new mutable.LinkedHashMap[String, BlockManagerId]()
```

Comment:
perhaps add a note saying both active and inactive and perhaps how things get removed when over max locations

Reply:
Maybe, we can add a shuffle service master role to manage the shuffle service worker.  App driver requests shuffle service workers from  shuffle service master. The master can  customize the politics to assign the workers for apps. Also, the master  manages the blacklist of shuffle service worker, adding or removing shuffle workers. In addition, the master can notify worker to delete data of completed apps by heart beat.